File: Fase3/aws-rekognition/FaceRokognitionJs/SubirColeccion.js
require("dotenv").config();
const fs = require("fs");
const {
  RekognitionClient,
  IndexFacesCommand,
} = require("@aws-sdk/client-rekognition");

const rekognitionClient = new RekognitionClient({
  region: process.env.AWS_REGION,
  credentials: {
    accessKeyId: process.env.AWS_ACCESS_KEY_ID,
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
  },
});

async function subirColeccion(nombreColeccion, path, externalImageId) {
  try {
    const bufferImg = fs.readFileSync(path);

    const command = new IndexFacesCommand({
      CollectionId: nombreColeccion,
      Image: {
        Bytes: bufferImg,
      },
      ExternalImageId: externalImageId,
      DetectionAttributes: ["ALL"],
    });

    const response = await rekognitionClient.send(command);

    console.log(
      `Rostro "${externalImageId}" subido a la colección "${nombreColeccion}" con éxito`
    );
    console.log(`ID asignado:", ${externalImageId}`);
    console.log("Detalles: ", response.FaceRecords[0].Face);

    return response;
  } catch (error) {
    console.error("Error al indexar el rostro ", error.name);
    if (error.name === "ResourceNotFoundException") {
      console.log("La colección no existe");
    }
    throw error;
  }
}

const [, , collectionName, path, externalImageId] = process.argv;
if (!collectionName || !path || !externalImageId) {
  console.log(
    "Faltan argumentos: nombre de la colección, ruta de la imagen y ID externo"
  );
  process.exit(1);
}

subirColeccion(collectionName, path, externalImageId);

File: Fase3/aws-rekognition/FRpiton/SubirColeccion.py
import os
import sys
import boto3
from dotenv import load_dotenv
from botocore.exceptions import BotoCoreError, ClientError

# Cargar variables de entorno
load_dotenv()

# Crear cliente de Rekognition
rekognition_client = boto3.client(
    'rekognition',
    region_name=os.getenv("AWS_REGION"),
    aws_access_key_id=os.getenv("AWS_ACCESS_KEY_ID"),
    aws_secret_access_key=os.getenv("AWS_SECRET_ACCESS_KEY")
)

def subir_coleccion(nombre_coleccion, path_imagen, external_image_id):
    try:
        with open(path_imagen, 'rb') as img_file:
            buffer_img = img_file.read()

        response = rekognition_client.index_faces(
            CollectionId=nombre_coleccion,
            Image={'Bytes': buffer_img},
            ExternalImageId=external_image_id,
            DetectionAttributes=['ALL']
        )

        print(f'Rostro "{external_image_id}" subido a la colección "{nombre_coleccion}" con éxito')
        if response['FaceRecords']:
            print("ID asignado:", external_image_id)
            print("Detalles:", response['FaceRecords'][0]['Face'])
        else:
            print("No se detectaron rostros.")

        return response
    except rekognition_client.exceptions.ResourceNotFoundException:
        print("La colección no existe")
    except (BotoCoreError, ClientError) as error:
        print("Error al indexar el rostro:", error)
        raise

# Leer argumentos desde la línea de comandos
if __name__ == "__main__":
    if len(sys.argv) < 4:
        print("Faltan argumentos: nombre de la colección, ruta de la imagen y ID externo")
        sys.exit(1)

    collection_name = sys.argv[1]
    path = sys.argv[2]
    external_image_id = sys.argv[3]

    subir_coleccion(collection_name, path, external_image_id)
